src/chat.ts

```diff
@@ -111,23 +118,38 @@ export class ChatProvider implements vscode.WebviewViewProvider {
     });
 
     webviewView.webview.onDidReceiveMessage(
-      async (
-        request:
-          | ChatRequest
-          | CancelRequest
-          | OpenLinkRequest
-          | InsertAtCursorInstruction
-      ) => {
+      async (request: ServerRequest | ExtensionRequest) => {
+        console.log(request);
         switch (request.type) {
-          case "chat_request": {
+          case "chatRequest": {
             vscode.commands.executeCommand("sourcery.chat_request", request);
             break;
           }
-          case "cancel_request": {
+          case "initialiseChat": {
+            vscode.commands.executeCommand("sourcery.initialise_chat");
+            break;
+          }
+          case "clearRequest": {
+            vscode.commands.executeCommand("sourcery.chat.clearChat");
+            break;
+          }
+          case "cancelRequest": {
             vscode.commands.executeCommand("sourcery.chat_cancel_request");
             break;
           }
-          case "open_link_request": {
+          case "recipeRequest": {
+            vscode.commands.executeCommand("sourcery.chat_request", request);
+            break;
+          }
```

Comment:
**🎯 suggestion:** ‏Remove all request and command specific code from src/chat.ts and src/extension.ts and move all logic into the webapp

*Goal* - make it so that when we add new messages from or to the webapp that we don't need to change the vscode (and therefore the pycharm) extensions. This has a few benefits:
- We only need to change the front and backend code, not the extensions. Right now the message also needs adding to chat.ts and extension.ts, as well as their PyCharm equivalents
- Additionaliy this means we only need to change a single repo to make a change, not three
- This will allow us to develop and test more quickly

One possible solution is that there's just one inbound message and one outbound message and these are routed directly through to the webapp / binary via the plugins. If the webapp needs extra info from the IDE it can use functions on the injected IDE interface.

Open to other solutions if you have them

*Note* - this can be done in follow up PRs

Reply:
Yeah I think this is going to be a good optimisation (I did more or less exactly your single-message suggestion for the troubleshooting view and it worked very well) and I'll add it as a change to make once the development has stabilised a bit. 

That said, I think there are actually two types of outbound message: ones for the server and ones for the extension (I've tried to codify this in the new type names). Neither is entirely transparent through the extension. The server ones will usually have additional data attached, such as the text selection, open files, and so on - this is information we don't have direct access to in the web app. We could pass it in to the web app (setting up relevant event listeners) or attach it on outbound messages - either seems reasonable. The extension ones will be used for e.g. the link handlers, direct text edits, and so on. We could skip these if we're happy for it to be a round trip to the server, but it's unclear exactly what the set of these IDE interactions is likely to be in the short or medium term, and routing through the server is getting a bit convoluted.

Basically all I'm saying is I don't want to jump too quickly into making this improvement before we have a better idea of the scope of it, but I agree with the principle.

Reply:
This is how I see where we should be aiming. To your point, there should be no direct app -> extension -> app loops.

![image](https://github.com/sourcery-ai/sourcery-vscode/assets/15631502/d3b123d1-d867-4cf3-b71a-2a2d7df2f27f)

Reply:
This all makes sense, and the distinction between server and extension requests makes a lot of sense

> The server ones will usually have additional data attached, such as the text selection, open files, and so on - this is information we don't have direct access to in the web app. We could pass it in to the web app (setting up relevant event listeners) or attach it on outbound messages

I think there's a third alternative which is probably cleaner. Rather than use event listeners to pass that information to the webapp, the webapp can request it from the extension through the injected IDE API

> Basically all I'm saying is I don't want to jump too quickly into making this improvement before we have a better idea of the scope of it, but I agree with the principle

Yep, fair enough. I would really like this before putting the webapp into PyCharm, so let's scope this out fully in our 1-1